### ISSUE TITLE
README.md Cons code only compiles with Prelude.Cons

### ISSUE THREAD

**Issue by khtan** [louthy/language-ext#601]
This is not really an issue but I am hoping to get some explanation.

In the README.md, there is a section on Lists that has an example :

   using LanguageExt;
   using static LanguageExt.Prelude;
   ...
   var test = Cons(1, Cons(2, Cons(3, Cons(4, Cons(5, empty<int>())))));

However, I found that for Visual Studio 2017, the compiler does not see Cons unless prefixed
with Prelude, ie

   var test = Prelude.Cons(1, Prelude.Cons(2, Prelude.Cons(3, Prelude.Cons(4, Prelude.Cons(5, empty<int>())))));

I looked at the git sources and also found that the tests such as LanguageExt.Tests/ArrayTests.cs and LanguageExt.Tests/ListTests.cs
also use Prelude.Cons instead of Cons.

It seems that Cons is an extension method and somehow, VS cannot find it unless the namespace is clearer, ie Prelude.Cons.

However, I am interested in a deeper or better explanation.

Thanks in advance,

**Comment by louthy** [comment]
Hi @khtan - there is some occasional weirdness with this for sure.  You should be able to call:
```c#
var test = 1.Cons(2.Cons(3.Cons(Empty))));
```

**Comment by louthy** [comment]
As for a deeper explanation I think it might have something to do with `Cons` taking a `this` parameter. I think when it's used directly in `static` include that C# *perhaps* ignores extension methods for direct use.  

I have occasionally bumped into this problem - and have promised myself to dig a bit deeper on a number of occasions.  Because it seems more like a compiler issue - but I'm loathe to ever claim a compiler issue without digging.

**Comment by khtan** [comment]
Thanks.

**Comment by khtan** [comment]
I think I found the reason. It is not a compiler bug. It is a "feature" of extension methods.

Cons is an extension method. See Prelude/Prelude_Collections.cs for some of the definitions :
   public static Seq<A> Cons<A>(this A ...)
   public static Lst<A> Cons<A>(this A ...)
   etc
Since Cons extends a type, constructs like 1.Cons compile correctly because the type of "1"
allows Cons to be found.

A little excerpt from "C# 7.0 in a Nutshell, Ch4 Dynamic Binding, Section Uncallable Functions"
provides the clarification.

   > Some functions cannot be called dynamically. You cannot call:
   > • Extension methods (via extension method syntax)
   > ...
   > Understanding why this is so is useful in understanding dynamic binding.
   > Dynamic binding requires two pieces of information: the name of the function to
   > call, and the object upon which to call the function. However, in each of the three
   > uncallable scenarios, an additional type is involved, which is known only at compile
   > time. As of C# 6, there’s no way to specify these additional types dynamically.
   > 
   > When calling extension methods, that additional type is implicit. It’s the static class
   > on which the extension method is defined. The compiler searches for it given the
   > using  directives in your source code. This makes extension methods compile-time-
   > only  concepts,  since  using   directives  melt  away  upon  compilation  (after  they’ve
   > done  their  job  in  the  binding  process  in  mapping  simple  names  to  namespace-
   > qualified names).

In other words, to use Cons, we need to use the namespace path, ie Prelude.Cons.
